Allow tuples to be converted to JsonNode by `%` in json.nim as discussed in #24082 .

siriuslee69 opened the pull request:
As discussed in issue #24082.

- Added the tuple type as a legit input type for `%`, to be able to convert named and unnamed tuples. Also changed part of the function itself to correctly discern between named and anonymous/unnamed tuples. The code was copied directly from jsonutils.nim (slightly changed).
- Additionally changed the openArray `%` proc to not allow tuples as array elements, such that 
 ```Nim
let 
    myArr = [someTuple1, someTuple2]
    myJsonNode = %myArr
 ```
cannot be compiled.
- Added / Copied the ``isNamedTuple()``  proc from the jsonutils.nim file instead of adding a dependency by using ``import std/typetraits`` .

- Added the respective tests in tjson.nim, both for the macro `%*` and the overloaded proc `%` for both anonymous/unnamed and named tuples.

- As in jsonutils.nim, the named tuples are converted to JObject while the anonymous tuples are converted to JArray.

Quick explanation as to why we had to disallow tuples for the openArray (so people don't have to scroll through the other convo):
Curly brackets together with a key value pair, kind of like this: `{"error": "No message"}` will be compiled to `[("error", "no message")]` by Nim. Which is an array of tuples. 
That means, if we do `%{"error": "No message"}`, the compiler will treat it as an array with a tuple inside. 
It will look like this under the hood: `%[("error", "No message")]`.
If we would pass this to the openarray function (with the new tuple `%` proc implemented), we should get:
`[["error", "No message"]]` .
And that is somehow a problem, idk why though tbh.

Related issues or issues that I copied code/comments from:
#16851, #12290, #16851, #10019, #10823

siriuslee69 committed:
allow tuples as input for `%`

Allow tuples as input for `%`, but disallow them in the generic type of the `%` proc that accepts openArrays.
Additionally addedthe isNamedTuple(T) proc. Can instead also be imported from std/typetraits, but ig dependencies should be kept to a minimum.

siriuslee69 committed:
added tests for tuple to JsObject / JsArray conversion

siriuslee69 committed:
added  some info the proc `%`

added additional conversion info to the proc, copied it straight from an earlier issue

siriuslee69 committed:
replaced toJson(v, opt) with %v

Forgot to change parts of the copied code from jsonutils.

siriuslee69 committed:
deleted unnecessary brackets in json.nim

siriuslee69 commented:
Ok, no clue what went wrong here, but apparently [this](https://github.com/nim-lang/Nim/blob/devel/tests/stdlib/tjsonmacro.nim#L555-L569) has some trouble with the new functions, if I understand the error messages correctly? What I find so weird is the fact that they apparently convert an array with a tuple with `%` ? But that shouldn't have worked before either? 

```Nim
import std/json

let
    myTestArrTup = @[(x: 123'u8, y: 53'u8, ch: 1231'u16)]
    myJsonNode = % myTestArrTup
#Doesn't compile
```

So why would this 
```Nim

  block:
    # bug related to #12015
    type
      PixInt = tuple[x, y, ch: int]
      SomePix = Pix | PixInt
      Cluster[T: SomePix] = seq[T]
      ClusterObject[T: SomePix] = object
        data: Cluster[T]
      RecoEvent[T: SomePix] = object
        cluster: seq[ClusterObject[T]]

    let data = @[(x: 123'u8, y: 53'u8, ch: 1231'u16)]
    var c = RecoEvent[Pix](cluster: @[ClusterObject[Pix](data: data)])
    let cFromJson = (% c).to(RecoEvent[Pix]) #<- why would this have worked before my commit? or did it not work before too?
    doAssert c == cFromJson

```

even be a valid test before? ```% c``` shouldn't work, or am I missing something? I don't know what Pix or Clusters are btw.

siriuslee69 commented:
I think [this](https://github.com/nim-lang/Nim/blob/devel/tests/stdlib/tjsonmacro.nim#L555-L569) is what causes the fails... but I don't really know why. Don't know what a Cluster or Pix is tbh. 
Maybe I made a mistake in the implementation somewhere? Idk. Will have to look through it tomorrow.

planetis-m commented:
There was an issue that would cause the generic overload with `tuple|object` to silently fail to apply. I don't remember any details sorry, but its related to: https://github.com/nim-lang/Nim/issues/16321

siriuslee69 committed:
split overloaded proc `%` for tuple and object in json.nim

Something like proc myproc[T: object|tuple](o:T): void = ...
might not accept the tuple properly even though it's supposed to, such that:
myproc(mytup) doesn't work. 
Idk if that was the actual problem here, but just to be sure, I changed the respective functions.
